### PR TITLE
Fixed #32556 -- Fixed handling empty string as non-boolean attributes value by assertHTMLEqual().

### PR DESCRIPTION
**Pull request by bmispelon**

### CODE REVIEW

File: docs/topics/testing/tools.txt [hunk 1]
```diff
@@ -1596,8 +1596,8 @@ your test suite.
       closed or the HTML document ends.
     * Empty tags are equivalent to their self-closing version.
     * The ordering of attributes of an HTML element is not significant.
-    * Attributes without an argument are equal to attributes that equal in
-      name and value (see the examples).
+    * `Boolean attributes`_ (like ``checked``, ``disabled``, ...) without an
+      argument are equal to attributes that equal in name and value (see the examples).
```
Comment (felixxm) [inline docs/topics/testing/tools.txt:1600]: I think we can chop the link:
```suggestion
    * Boolean attributes (like ``checked``) without an argument are equal to
      attributes that equal in name and value (see the examples).
```
Also, `versionchanged` annotation seems appropriate, e.g.
```
.. versionchanged:: 4.0

    In older versions, non-boolean attributes without an argument were equal to
    attributes that equal in name and value.
```

File: docs/releases/4.0.txt [hunk 2]
```diff
@@ -261,6 +261,12 @@ Tests
 * Django test runner now supports a :option:`--buffer <test --buffer>` option
   with parallel tests.
 
+* :meth:`~django.test.SimpleTestCase.assertHTMLEqual` now handles
+  `boolean attributes <https://html.spec.whatwg.org/#boolean-attribute>`_
+  more correctly so it knows that ``<input checked="">`` is the same as
+  ``<input checked="checked">`` but that ``<input value="">`` is not the same
+  as ``<input value="value">``.
```
Comment (felixxm) [inline docs/releases/4.0.txt:268]: I don't think we need release notes, maybe a small note in the ["Backwards incompatible changes in 4.0"](https://docs.djangoproject.com/en/dev/releases/4.0/#miscellaneous) section.

**Review by felixxm — COMMENTED** [review]
@bmispelon Thanks for this patch :+1: I left small comments.

File: docs/topics/testing/tools.txt (hunk 1 above)
Comment (bmispelon) [reply docs/topics/testing/tools.txt]: I tweaked the `versionchanged` text a bit from what you suggested.

File: django/test/html.py [hunk 3]
```diff
@@ -53,24 +61,7 @@ def rstrip_last_element(children):
...
-                attr, value = self.attributes[i]
-                other_attr, other_value = element.attributes[i]
-                if not value:
-                    value = attr
-                if not other_value:
-                    other_value = other_attr
-                if attr != other_attr or value != other_value:
-                    return False
-        return self.children == element.children
+        return str(self) == str(element)
```
[the remark this answers is not in the document]
Comment (bmispelon) [reply django/test/html.py]: I added a 3rd commit with a different approach that keeps the separate implementation of `__eq__` and `__str__` but normalizes the `attrs` in the parser.

**Comment by felixxm** [comment]
@bmispelon Thanks for updates :+1:

I moved adding `normalize_attributes()` to a separate commit, and added assertions for the string representation.  I'll check the docs changes tonight.

**Comment by felixxm** [comment]
@bmispelon Thanks again :+1: 

I pushed small edits to docs.

**Comment by bmispelon** [comment]
Nicely done, thanks! ✨

**Pull request merged**